zrml/neo-swaps/src/lib.rs

```diff
@@ -0,0 +1,903 @@
+                    Error::<T>::Unexpected
+                );
+                ensure!(amount_in <= pool.calculate_max_amount_in(), Error::<T>::NumericalLimits);
+                // Instead of first executing a swap with `(n-1)` transfers from the pool account to
+                // `who` and then selling complete sets, we prevent `(n-1)` storage reads: 1)
+                // Transfer `amount_in` units of `asset_in` to the pool account, 2) sell
+                // `amount_out` complete sets using the pool account, 3) transfer
+                // `amount_out_minus_fees` units of collateral to `who`. The fees automatically end
+                // up in the pool.
+                let amount_out = pool.calculate_swap_amount_out_for_sell(asset_in, amount_in)?;
+                // Beware! This transfer happen _after_ calculating `amount_out`:
+                T::AssetManager::transfer(asset_in, &who, &pool.account_id, amount_in)?;
+                T::CompleteSetOperations::sell_complete_set(
+                    pool.account_id.clone(),
+                    market_id,
+                    amount_out,
+                )?;
+                let FeeDistribution {
+                    remaining: amount_out_minus_fees,
+                    swap_fees: swap_fee_amount,
+                    external_fees: external_fee_amount,
+                } = Self::distribute_fees(market_id, pool, amount_out)?;
+                ensure!(amount_out_minus_fees >= min_amount_out, Error::<T>::AmountOutBelowMin);
+                T::AssetManager::transfer(
+                    pool.collateral,
+                    &pool.account_id,
+                    &who,
+                    amount_out_minus_fees,
+                )?;
+                for (&asset, balance) in pool.reserves.iter_mut() {
+                    if asset == asset_in {
+                        *balance = balance.checked_add(&amount_in).ok_or(Error::<T>::MathError)?;
+                    }
+                    *balance = balance.checked_sub(&amount_out).ok_or(Error::<T>::MathError)?;
+                }
+                let new_price = pool.calculate_spot_price(asset_in)?;
+                ensure!(
+                    new_price >= MIN_SPOT_PRICE.saturated_into(),
+                    Error::<T>::SpotPriceBelowMin
+                );
+                Self::deposit_event(Event::<T>::SellExecuted {
+                    who: who.clone(),
+                    market_id,
+                    asset_in,
+                    amount_in,
+                    amount_out,
+                    swap_fee_amount,
+                    external_fee_amount,
+                });
+                Ok(())
+            })
+        }
+
+        #[require_transactional]
+        fn do_join(
+            who: T::AccountId,
+            market_id: MarketIdOf<T>,
+            pool_shares_amount: BalanceOf<T>,
+            max_amounts_in: Vec<BalanceOf<T>>,
```

Comment:
Maybe a key-value map would be more intuitive and less error-prone here.

Reply:
I would ask to make this a task for the full implementation. We're kind of in a hurry here and I want the good people of the Frontend Team to have as little work as possible, so would prefer to keep the old interface.

I'm with you though, I'm slightly creeped out by how much I'm relying on the `assets` object being ordered.

Reply:
https://github.com/zeitgeistpm/zeitgeist/issues/1107